### PR TITLE
Adds bump attack component to humans, adds bump attack compatibility to harvester and unarmed

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request
Adds bump attacks to harvester and fists.

You still can't FF with bump attacks or bump attack with a gun.

Tested bump attack component changes with a runner and queen and they should work the same as before.

## Why It's Good For The Game
Update: The bump attacks with wielded requirement and damage output reduction were tested and it appears that due to marines moving slower than xenos, especially when damaged or wearing heavy armor, bump attacks aren't as reliable as they are on xenos. This is due to them happening when you move and once a marine moves, it is unable to try and hit something by using the movement keys before it can move again. Unless marines are moving nearly as fast as xenos, this doesn't seem to be as big of a buff as it sounds for this weapon.

Targeting with melee is very clunky and this makes attacking more consistent, while also decreasing potential damage. This should give the option to attack in melee with less, but consistent, DPS rather than deal too much damage when your hits connect or none when they don't.

I think this would provide the option for having more consistent performance as a marine with a lower skill floor in exchange for decreased peak damage, which would help with increasing marines' overall ability to hold an advance, rather than rely too much on who's playing at the moment.

## Changelog
:cl:
add: added bump attack component to humans, added bump attack compatibility to wielded harvester and unarmed
/:cl:

### CODE REVIEW

File: code/game/objects/items/weapons/blades.dm [hunk 1]
```diff
@@ -28,8 +28,9 @@
 	force = 60
 	attack_speed = 12
 	w_class = WEIGHT_CLASS_BULKY
-	flags_item = DRAINS_XENO
+	flags_item = DRAINS_XENO|TWOHANDED
```
Comment (BraveMole) [inline code/game/objects/items/weapons/blades.dm:31]: undocumented? 
Also, wasn't vali only slighlty viable with a shield? I don't think bumpslash will make it sudendly OP

**Comment by MLGTASTICa** [comment]
imagine if this was added onto shotguns..

File: code/datums/components/bump_attack.dm [hunk 2]
```diff
@@ -2,34 +2,38 @@
 	var/active = TRUE
 	var/bump_action_path
 	var/datum/action/bump_attack_toggle/toggle_action
+	var/mob/living/bumper
```
Comment (BraveMole) [inline code/datums/components/bump_attack.dm:5]: You're not supposed to hold a casted reference to a component parent, because this removes some genericty.
That being said, are bumped attacked supposed to be used by something not a mob/living? I don't think it will ever be the case, so why no. Not sure Tivi will like this though

File: code/datums/components/bump_attack.dm [hunk 3]
```diff
@@ -69,33 +69,47 @@
 
 /datum/component/bump_attack/proc/human_bump_action(datum/source, atom/target)
 	SIGNAL_HANDLER
-	var/mob/living/carbon/human/bumper = parent
+	INVOKE_ASYNC(src, .proc/human_bump_action_checks, source, target)
```
Comment (BraveMole) [inline code/datums/components/bump_attack.dm:72]: I'm not seing any sleep in the next procs, why are you using a wrapper function with an INVOKE ASYNC?

File: code/modules/mob/living/carbon/human/human.dm [hunk 4]
```diff
@@ -41,6 +41,7 @@
 	RegisterSignal(src, COMSIG_KB_RAILATTACHMENT, .proc/do_activate_rail_attachment)
 	RegisterSignal(src, COMSIG_GRAB_SELF_ATTACK, .proc/fireman_carry_grabbed) // Fireman carry
 	AddComponent(/datum/component/footstep, FOOTSTEP_MOB_HUMAN)
+	AddComponent(/datum/component/bump_attack)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/human/human.dm:44]: I don't like very much that approach. That means that every human will get the toggle action, while less than 2-5% will actually use it. We don't want to have useless actions lying around.

Even performance wise, you are checking if human can bump or not, even if they are using a minigun loadout... Doesn't make sense.

Either you manage to make the giving action and the registering only when you wield the vali harvester, or you make a bump component for the weapon and not for the living. Also, who cares about unarmed attacks, don't tackle that unless it's really easy to add and doesn't lead to shitcode

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/datums/components/bump_attack.dm [hunk 5]
```diff
@@ -2,34 +2,38 @@
...
+	toggle_action = new()
+	toggle_action.give_action(bumper)
 	toggle_action.update_button_icon(active)
 	RegisterSignal(toggle_action, COMSIG_ACTION_TRIGGER, toggle_path)
-	if(active)
-		RegisterSignal(parent, COMSIG_MOVABLE_BUMP, bump_action_path)
 
 /datum/component/bump_attack/Destroy(force, silent)
+	if(active)
+		UnregisterSignal(bumper, COMSIG_MOVABLE_BUMP)
```
Comment (BraveMole) [inline code/datums/components/bump_attack.dm:30]: unneeded, signals are automaticly cleaned on object destroy

File: code/datums/components/bump_attack.dm [hunk 6]
```diff
@@ -2,34 +2,38 @@
...
 	QDEL_NULL(toggle_action)
+	bumper = null
```
Comment (BraveMole) [inline code/datums/components/bump_attack.dm:32]: Don't take my word for it, but i think you also need to clean that ref when bumper is QDEL. Ask tivi about this, I don't think you are protecting this from hard del enough. It's possible that the component is automaticly destroyed on the parent destroy.

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/blades.dm (hunk 1 above)
Comment (Walarks) [reply code/game/objects/items/weapons/blades.dm:31]: This is required to allow harvester to be wielded.

File: code/datums/components/bump_attack.dm (hunk 2 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm:5]: I think the core signals and checks used for the component require it to be on a living being, so if it was used for object, it would need to be rewritten anyways. 

I added this because for some reason "var/mob/living/bumper = parent" on each proc caused runtimes when using the component on humans.

File: code/game/objects/items/weapons/blades.dm (hunk 1 above)
Comment (BraveMole) [reply code/game/objects/items/weapons/blades.dm:31]: Why are you requiring it to be wielded for the bump attack? Is it a design intent or because you didn't find a way to do it without?

Comment (Walarks) [reply code/game/objects/items/weapons/blades.dm:31]: It's a design intent so people don't use shield + vali and such.

Comment (BraveMole) [reply code/game/objects/items/weapons/blades.dm:31]: Eh, your choice, I don't see why you keep making vali a weak setup

**Comment by Losenis** [comment]
machetes get this too?

**Comment by Walarks** [comment]
> machetes get this too?

No, probably after this has been tested and is ok to be a thing.

File: code/datums/components/bump_attack.dm (hunk 2 above)
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:5]: this is terribad

File: code/datums/components/bump_attack.dm (hunk 5 above)
Comment (TiviPlus) [reply code/datums/components/bump_attack.dm:30]: yes

File: code/datums/components/bump_attack.dm [hunk 7]
```diff
@@ -69,33 +69,47 @@
...
+
+/datum/component/bump_attack/proc/human_bump_action_checks(datum/source, atom/target)
+	var/mob/living/carbon/human/attacker = bumper
 	. = carbon_bump_action_checks(target)
 	if(!isnull(.))
 		return
 	var/mob/living/living_target = target
-	if(bumper.faction == living_target.faction)
+	if(attacker.faction == living_target.faction)
 		return //FF
```
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:81]: ??

File: code/datums/components/bump_attack.dm [hunk 8]
```diff
@@ -69,33 +69,47 @@
...
 	if(bumper.next_move > world.time)
 		return COMPONENT_BUMP_RESOLVED //We don't want to push people while on attack cooldown.
 	bumper.UnarmedAttack(target, TRUE)
 	GLOB.round_statistics.xeno_bump_attacks++
 	SSblackbox.record_feedback("tally", "round_statistics", 1, "xeno_bump_attacks")
 	TIMER_COOLDOWN_START(src, COOLDOWN_BUMP_ATTACK, CLICK_CD_MELEE)
 	return COMPONENT_BUMP_RESOLVED
+
+/datum/component/bump_attack/proc/human_do_bump_action(atom/target)
+	if(bumper.next_move > world.time)
```
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:105]: No autodoc

File: code/game/objects/items/weapons/blades.dm (hunk 1 above)
Comment (TiviPlus) [inline code/game/objects/items/weapons/blades.dm:31]: hello yes unrelated changes

File: code/modules/mob/living/carbon/human/human.dm (hunk 4 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/human.dm:44]: yes that

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/blades.dm (hunk 1 above)
Comment (Walarks) [reply code/game/objects/items/weapons/blades.dm:31]: It needs the twohanded flag to be able to be wielded.

**Comment by bowlofcereal** [comment]
wtf no bump to machete but bump to harvester? this is a personal attack

File: code/datums/components/bump_attack.dm (hunk 8 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: autodocced

File: code/datums/components/bump_attack.dm (hunk 7 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: removed var, INVOKE_ASYNC is needed for sleeping procs

File: code/datums/components/bump_attack.dm (hunk 2 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: removed var

File: code/datums/components/bump_attack.dm (hunk 6 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: removed var

File: code/datums/components/bump_attack.dm (hunk 5 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: removed section

File: code/modules/mob/living/carbon/human/human.dm (hunk 4 above)
Comment (Walarks) [reply code/modules/mob/living/carbon/human/human.dm]: Made it so that the component's checks and such need to be enabled for the component to run.

File: code/datums/components/bump_attack.dm (hunk 3 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm:75]: There are sleeps in procs that are called further down.

File: code/datums/components/bump_attack.dm (hunk 2 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: removed var

**Comment by Smudged** [comment]
Bump attack to FIST?! 

This is a mistake, of well, just marines walking forward

**Comment by davidchan13** [comment]
Gimme Bayonet and Entrenching Tool bump attacks.

File: code/datums/components/bump_attack.dm [hunk 9]
```diff
@@ -1,42 +1,51 @@
 /datum/component/bump_attack
-	var/active = TRUE
+	///Whether the component is active
+	var/active
```
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:3]: Set the default for bools to be = FALSE

File: code/datums/components/bump_attack.dm [hunk 10]
```diff
@@ -1,42 +1,51 @@
...
-/datum/component/bump_attack/proc/living_activation_toggle(datum/source)
+/datum/component/bump_attack/proc/living_activation_toggle(datum/source, should_enable = !active)
+	if(should_enable == active)
+		return
 	var/mob/living/bumper = parent
-	active = !active
 	to_chat(bumper, "<span class='notice'>You will now [active ? "attack" : "push"] enemies who are in your way.</span>")
-	if(active)
+	if(toggle_action)
+		toggle_action.update_button_icon(should_enable)
```
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:38]: use .?

File: code/datums/components/bump_attack.dm [hunk 11]
```diff
@@ -1,42 +1,51 @@
...
+	if(should_enable)
+		active = TRUE
 		RegisterSignal(bumper, COMSIG_MOVABLE_BUMP, bump_action_path)
-	else
-		UnregisterSignal(bumper, COMSIG_MOVABLE_BUMP)
-	toggle_action.update_button_icon(active)
+		return
+
+	var/obj/item/held_item = bumper.get_inactive_held_item()
+	if(held_item && held_item.flags_item & CAN_BUMP_ATTACK)
```
Comment (TiviPlus) [inline code/datums/components/bump_attack.dm:45]: Use .?

File: code/datums/components/bump_attack.dm (hunk 3 above)
Comment (TiviPlus) [reply code/datums/components/bump_attack.dm:78]: Then it should only invoke the proc that actually sleeps which I assume is human_do_bump_action

File: code/game/objects/items/weapons/blades.dm [hunk 12]
```diff
@@ -28,8 +28,9 @@
...
 
+	var/force_wielded = 40
```
Comment (TiviPlus) [inline code/game/objects/items/weapons/blades.dm:33]: Autodoc

File: code/game/objects/items.dm [hunk 13]
```diff
@@ -355,6 +355,16 @@
...
 
+///Used to enable/disable an item's bump attack. Grouped in a proc to make sure the signal or flags aren't missed
+/obj/item/proc/toggle_item_bump_attack(mob/user, enable_bump_attack)
+	SEND_SIGNAL(user, COMSIG_ITEM_TOGGLE_BUMP_ATTACK, enable_bump_attack)
+	if(flags_item & CAN_BUMP_ATTACK && enable_bump_attack)
+		return
+	if(enable_bump_attack)
+		flags_item |= CAN_BUMP_ATTACK
+		return
+	flags_item &= ~CAN_BUMP_ATTACK
```
Comment (TiviPlus) [inline code/game/objects/items.dm:366]: Why is this proc neeeded

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/datums/components/bump_attack.dm (hunk 9 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm:3]: changed

File: code/datums/components/bump_attack.dm (hunk 10 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm:38]: changed

File: code/datums/components/bump_attack.dm (hunk 11 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm:45]: changed

File: code/game/objects/items/weapons/blades.dm (hunk 12 above)
Comment (Walarks) [reply code/game/objects/items/weapons/blades.dm:33]: changed

File: code/game/objects/items.dm (hunk 13 above)
Comment (Walarks) [reply code/game/objects/items.dm:366]: It's needed since both the bitflag and signal are necessary for the component to function properly and this makes the code more ordered.

File: code/datums/components/bump_attack.dm (hunk 11 above)
Comment (Walarks) [reply code/datums/components/bump_attack.dm]: changed

**Review by TiviPlus — APPROVED** [review]
Tested?

**Comment by Walarks** [comment]
> Tested?

Works based on my tests

**Review by TiviPlus — APPROVED** [review]
